### ISSUE TITLE
Mangling _FloatN types

### ISSUE THREAD

**Issue by smeijer1234** [itanium-cxx-abi/cxx-abi#21]
In C11 extension ISO/IEC TS 18661, new floating point extensions and types are defined. For example, _FloatN is introduced as a binary interchange format, where N can be 16, 32, 64, 128 (or bigger). We are implementing _Float16 support in Clang and also like to use it in C++ mode. Therefore we need mangling support for _Float16. Is this something the C++ ABI would consider to support?

**Comment by thiagomacieira** [comment]
Is _Float16 the same as __fp16, which already supported on some architectures? If so, should the two types be the same?

And if so, please remember there may already be code that uses __fp16 in mangled contexts, using mangling "Dh".
```
$ clang -target aarch64-linux -O2 -S -o - -xc++ - <<<'void f(const __fp16 &) {}' 
        .text
        .file   "-"
        .globl  _Z1fRKDh                // -- Begin function _Z1fRKDh
        .p2align        2
        .type   _Z1fRKDh,@function
_Z1fRKDh:                               // @_Z1fRKDh
// BB#0:
        ret
.Lfunc_end0:
        .size   _Z1fRKDh, .Lfunc_end0-_Z1fRKDh
                                        // -- End function

        .ident  "openSUSE Linux clang version 5.0.0 (trunk 307005) (based on LLVM 5.0.0)"
        .section        ".note.GNU-stack","",@progbits
```

PS: __float128 is also already mangled as "g"

**Comment by smeijer1234** [comment]
No, they are really different types: __fp16 is a storage-only type (arithmetic is performed in single precision, the result truncated and stored in half precision), and _Float16 is a "native" type with arithmetic in half precision. As they are different types with different semantics, they should be mangled differently.

**Comment by rjmccall** [comment]
If TS 18661 considers them different types, then the C++ ABI should provide new manglings for them unless we have some compelling reason to reject using that extension in C++, which seems unlikely.

I recommend:
  <builtin-type> ::= 'DF' <number>

Would you mind submitting a pull request?

**Comment by smeijer1234** [comment]
Many thanks and yes, I will start preparing a patch.

**Comment by rjmccall** [comment]
That patch is #21, for the record.

**Comment by rjmccall** [comment]
Patch approved.  Thanks!

**Comment by rjmccall** [comment]
@smeijer1234 Are you aware of any implementations that currently use this mangling?
